File: ui/chat/src/chat-status.ts
import type { ChatStatus } from "./chat-panel-types";
import type { FeedItem } from "./types";

/**
 * Derive the chat-panel status from the feed and the controller's loading
 * flag. The status decides whether [`ChatMessages`](./chat-messages.tsx)
 * renders the in-flight loading indicator (`"submitted"`) or hides it
 * because the actual stream is the progress signal (`"streaming"`).
 *
 * Only `assistant_text_streaming` counts as "streaming": it is the one
 * feed-item type whose progressively-appearing content is VISIBLE, so the
 * indicator would just compete with it. `thinking_streaming` used to count
 * too, but since HOU-448 the reasoning streams inside the collapsed-by-
 * default mission log — nothing on screen moves, so treating it as
 * "streaming" made the loading helmet flicker off during every thinking
 * stretch (HOU-655 follow-up). EVERY case where a turn is in flight and no
 * visible text is streaming resolves to "submitted" so the user sees a
 * loading indicator.
 *
 * The previous logic returned `"streaming"` whenever `isLoading` was
 * true and the chat had any prior items — which hid the indicator
 * during the multi-second silent stretches that Gemini introduces (it
 * emits an init line, optionally fires its auto `update_topic` tool,
 * then sits silent for 10-20s before bursting the entire response in
 * one batch). The indicator is the only signal during those stretches.
 */
export function deriveStatus(
  items: FeedItem[],
  isLoading: boolean,
): ChatStatus {
  const last = items[items.length - 1];
  if (last?.feed_type === "assistant_text_streaming") {
    return "streaming";
  }
  // A typed provider error is a TERMINAL settle — the agent cannot continue
  // (out of usage, rate-limited, logged out). Whatever a stale `isLoading`
  // flag says (a lost terminal frame can leave the VM's running flag up until
  // the idle reconcile), the honest state is "the turn is over": no rotating
  // thinking phrases under a card that says the work stopped (PRODUCT-1578).
  // The card's trailing invisible final_result (and any session-error echo)
  // is skipped when locating the transcript's last visible event.
  if (endsOnProviderError(items)) return "ready";
  // Active turn → indicator visible. Covers:
  //   - brand-new chat with no items yet
  //   - user just sent (last == user_message)
  //   - provider mid-tool-cycle (last == tool_call / tool_result),
  //     waiting on the model's next chunk
  //   - reasoning streaming or just landed inside the collapsed
  //     mission log, still waiting on the response
  //   - any silent gap between tokens for batchy providers (Gemini)
  if (isLoading) return "submitted";
  // Idle but the user's message is the last thing on the feed — either the
  // send just fired and `isLoading` flips true on the next tick, or the
  // message is parked while the agent's engine warms up (HOU-693/713).
  // Both are "a reply is coming": treat as in-flight.
  if (last?.feed_type === "user_message") {
    return "submitted";
  }
  return "ready";
}

/**
 * Whether the feed's last VISIBLE event is a terminal `provider_error` card.
 * `final_result` (the settle's invisible frame) and `system_message` (the
 * session-status echo a failed turn can trail) are skipped; anything else —
 * a user message, streaming content, a tool row — means the conversation
 * moved past the failure.
 */
function endsOnProviderError(items: FeedItem[]): boolean {
  for (let i = items.length - 1; i >= 0; i--) {
    const type = items[i].feed_type;
    if (type === "final_result" || type === "system_message") continue;
    return type === "provider_error";
  }
  return false;
}

File: mobile/ios/TilinX/Features/Chat/ChatStatus.swift
import Foundation

/// The chat's in-flight display state, mirroring the desktop `ChatStatus`
/// (`ui/chat/src/chat-panel-types.ts`) and its `deriveStatus` reclassification
/// (`ui/chat/src/chat-status.ts`).
///
/// Only `assistant_text_streaming` counts as `.streaming`: it is the ONE feed
/// type whose progressively-appearing content is visible on screen, so the
/// loading indicator would just compete with it. `thinking_streaming` used to
/// count too, but since HOU-448 the reasoning streams inside a collapsed block —
/// nothing moves on screen — so treating it as streaming flickered the loading
/// state off during every thinking stretch (HOU-655). EVERY in-flight case with
/// no visible text streaming resolves to `.submitted` so the loading indicator
/// stays visible through reasoning + tool phases.
enum ChatStatus: Equatable, Sendable {
  /// Settled — no loading indicator.
  case ready
  /// Assistant text is streaming; the bubble IS the progress signal, so the
  /// standalone loading indicator is suppressed.
  case streaming
  /// A turn is in flight with nothing visibly streaming — show the loading
  /// indicator (the only signal during reasoning / tool / silent-gap stretches).
  case submitted

  /// Reclassify the feed + running flag into a status, byte-for-byte the desktop
  /// `deriveStatus(items, isLoading)`: last frame `assistant_text_streaming` →
  /// `.streaming`; otherwise any running turn (or a just-sent optimistic
  /// `user_message`) → `.submitted`; else `.ready`.
  static func derive(feed: [FeedItemVM], running: Bool) -> ChatStatus {
    let lastType = feed.last?.feedType
    if lastType == "assistant_text_streaming" { return .streaming }
    if running { return .submitted }
    if lastType == "user_message" { return .submitted }
    return .ready
  }
}
